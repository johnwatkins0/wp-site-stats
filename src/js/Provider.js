/* eslint camelcase: 0 */
import React from 'react';
import get from 'lodash.get';

import SiteStats from './SiteStats';

import { setActiveSiteId, setRefreshTimer } from './store/actions';
import effects from './store/effects';
import reducer from './store/reducer';
import { getSecondsToRefresh } from './store/selectors';
import { REFRESH_SECONDS } from './constants';
import { AppContext, ActiveSiteContext } from './Context';

const DEFAULT_SITE_DATA = {
	site_title: '',
	site_url: '',
	latest_post: null,
	latest_comment: null
};

class Provider extends React.Component {
	constructor( props ) {
		super( props );
		this.state = {
			app: {
				sites: props.sites,
				secondsToRefresh: REFRESH_SECONDS,
				activeSiteId: '0',
				shouldRefresh: true
			},
			siteData: {
				'0': DEFAULT_SITE_DATA
			}
		};
		this.getState = this.getState.bind( this );
		this.dispatch = this.dispatch.bind( this );
		this.setActiveSiteId = this.setActiveSiteId.bind( this );
		this.tick = this.tick.bind( this );
	}

	componentDidMount() {
		this.dispatch( setActiveSiteId( this.props.sites[0].id ) );
		this.startTicker();
	}

	componentWillUpdate( prevProps, nextState ) {
		this.maybeStopTicker( nextState );
	}

	getState() {
		return Object.assign({}, this.state );
	}

	setActiveSiteId( siteId ) {
		this.dispatch( setActiveSiteId( siteId ) );
	}

	dispatch( action )  {
		this.setState(
			( state ) => reducer( state, action ),
			() => this.triggerEffect( action )
		);
	}

	triggerEffect( action ) {
		if ( action.type in effects ) {
			effects[action.type](
				action,
				{
					dispatch: this.dispatch,
					getState: this.getState
				}
			);
		}
	}

	startTicker() {
		this.ticker = setInterval( this.tick, 1000 );
	}

	tick() {
		this.dispatch(
			setRefreshTimer( getSecondsToRefresh( this.getState() ) - 1 )
		);
	}

	maybeStopTicker( nextState ) {
		if ( false !== this.props.shouldRefresh && false === nextState.shouldRefresh ) {
			clearInterval( this.ticker );
		}
	}

	render() {
		return (
			<AppContext.Provider value={ this.state.app }>
				<ActiveSiteContext.Provider
					value={ this.state.siteData[this.state.app.activeSiteId] ||
						DEFAULT_SITE_DATA }
				>
					<SiteStats setActiveSite={this.setActiveSiteId} />
				</ActiveSiteContext.Provider>
			</AppContext.Provider>
		);
	}
}

export default Provider;
