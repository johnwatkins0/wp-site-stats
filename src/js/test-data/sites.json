[
    {
        "id": "1",
        "domain": "wpmu",
        "path": "/",
        "registered": "2018-04-02 14:28:19"
    },
    {
        "id": "2",
        "domain": "wpmu",
        "path": "/my-site/",
        "registered": "2018-04-02 14:29:11"
    },
    {
        "id": "3",
        "domain": "wpmu",
        "path": "/another-site/",
        "registered": "2018-04-02 14:29:27"
    },
    {
        "id": "4",
        "domain": "wpmu",
        "path": "/cool-site/",
        "registered": "2018-04-03 17:13:22"
    },
    {
        "id": "5",
        "domain": "wpmu",
        "path": "/another-cool-site/",
        "registered": "2018-04-03 17:13:32"
    },
    {
        "id": "6",
        "domain": "wpmu",
        "path": "/better-site/",
        "registered": "2018-04-03 17:13:44"
    },
    {
        "id": "7",
        "domain": "wpmu",
        "path": "/best-site/",
        "registered": "2018-04-03 17:13:53"
    }
]