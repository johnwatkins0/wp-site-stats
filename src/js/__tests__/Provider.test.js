import React from 'react';
import renderer from 'react-test-renderer';
import Enzyme, { shallow } from 'enzyme';
import Adapter from 'enzyme-adapter-react-16';

import Provider from '../Provider';

import sites from '../test-data/sites.json';

Enzyme.configure({adapter: new Adapter() });

describe( 'SiteSelector', () => {
    it( 'renders', () => {
        const tree = renderer.create(
            <Provider
                sites={sites}
            />
        ).toJSON();
        expect( tree ).toMatchSnapshot();
    });

    it( 'starts ticking', ( done ) => {
        const wrapper = shallow(
            <Provider
                sites={sites}
            />
        );

        setTimeout( () => {
            expect( wrapper.state().secondsToRefresh ).toEqual( 59 );
            done();
        }, 1001 );
    });

    it( 'can stop ticking', ( done ) => {
        const wrapper = shallow(
            <Provider
                sites={sites}
            />
        );

        wrapper.setState({ shouldRefresh: false });

        setTimeout( () => {
            expect( wrapper.state().secondsToRefresh ).toEqual( 60 );
            done();
        }, 1001 );
    });
})
;
