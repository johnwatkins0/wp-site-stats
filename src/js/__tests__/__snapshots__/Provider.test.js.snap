// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SiteSelector renders 1`] = `
<div
  className="SiteStats"
>
  <div
    className="SiteStats__header"
  >
    <h1>
      Site Stats
    </h1>
    <label>
      Sites:
       
      <select
        className="SiteSelector"
        onChange={[Function]}
        value="1"
      >
        <option
          value="1"
        >
          wpmu/
        </option>
        <option
          value="2"
        >
          wpmu/my-site/
        </option>
        <option
          value="3"
        >
          wpmu/another-site/
        </option>
        <option
          value="4"
        >
          wpmu/cool-site/
        </option>
        <option
          value="5"
        >
          wpmu/another-cool-site/
        </option>
        <option
          value="6"
        >
          wpmu/better-site/
        </option>
        <option
          value="7"
        >
          wpmu/best-site/
        </option>
      </select>
    </label>
  </div>
  <div
    className="SiteStats__body"
  >
    <article
      className="MainPane"
    >
      <header
        className="MainPane__header"
      >
        <h2
          dangerouslySetInnerHTML={
            Object {
              "__html": "&nbsp;",
            }
          }
        />
        <div
          className="MainPane__site-info"
        >
          <div>
            <span
              className="MainPane__site-info-key"
            >
              URL:
            </span>
             
            <a
              href={undefined}
            />
          </div>
          <div>
            <span
              className="MainPane__site-info-key"
            >
              Created:
            </span>
             
            April 2, 2018
          </div>
        </div>
      </header>
      <section
        className="MainPane__section"
      >
        <h3>
          By the Numbers
        </h3>
        <div
          className="MainPane__stats"
        >
          <div
            className="MainPane__stat-container"
          >
            <div
              className="MainPane__stat"
            >
              <h4
                className="MainPane__stat-title"
              >
                Posts
              </h4>
              <div
                className="MainPane__stat-number"
              >
                -
              </div>
            </div>
          </div>
          <div
            className="MainPane__stat-container"
          >
            <div
              className="MainPane__stat"
            >
              <h4
                className="MainPane__stat-title"
              >
                Pages
              </h4>
              <div
                className="MainPane__stat-number"
              >
                -
              </div>
            </div>
          </div>
          <div
            className="MainPane__stat-container"
          >
            <div
              className="MainPane__stat"
            >
              <h4
                className="MainPane__stat-title"
              >
                Users
              </h4>
              <div
                className="MainPane__stat-number"
              >
                -
              </div>
            </div>
          </div>
          <div
            className="MainPane__stat-container"
          >
            <div
              className="MainPane__stat"
            >
              <h4
                className="MainPane__stat-title"
              >
                Comments
              </h4>
              <div
                className="MainPane__stat-number"
              >
                -
              </div>
            </div>
          </div>
        </div>
      </section>
      <section
        className="MainPane__section MainPane__latest-post"
      >
        <h3>
          Latest Post
        </h3>
        <div
          className="MainPane__latest-post-link"
        >
          <a
            dangerouslySetInnerHTML={
              Object {
                "__html": "(no title)",
              }
            }
            href={undefined}
          />
        </div>
      </section>
      <section
        className="MainPane__section MainPane__latest-comment"
      >
        <h3>
          Latest Comment
        </h3>
        <div
          className="MainPane__latest-comment-link"
        >
          <a
            href={undefined}
          >
             
					on Invalid Date
					at Invalid Date
          </a>
        </div>
      </section>
      <footer
        className="MainPane__footer"
      >
        Refreshing in 
        60
         second
        s
        .
      </footer>
    </article>
  </div>
</div>
`;
